File: cpp/triangle/triangle.cpp
#include "triangle.h"
#include <stdexcept>

namespace triangle {
    flavor kind(double a, double b, double c) 
    {
        if (a <= 0.0 || b <= 0.0 || c <= 0.0)
            throw std::domain_error("Invalid input: All values must be higher than zero");
        
        if (!satisfies_triangle_inequality(a, b, c))
            throw std::domain_error("Invalid input: Values doesn't satisfy triangle inequality");
        
        if (is_equilateral(a, b, c))
            return flavor::equilateral;
        else if (is_isosceles(a, b, c))
            return flavor::isosceles;
        else
            return flavor::scalene;
    }

    bool satisfies_triangle_inequality(double a, double b, double c)
    {
        return a + b >= c && a + c >= b && b + c >= a;
    }

    bool is_equilateral(double a, double b, double c)
    {
        return a == b && b == c;
    }

    bool is_isosceles(double a, double b, double c)
    {
        return a == b || b == c || a == c;
    }
}  // namespace triangle

File: prolog/triangle/triangle.pl
triangle(A, B, C, _) :- A + B < C, !, fail.
triangle(A, B, C, _) :- B + C < A, !, fail.
triangle(A, B, C, _) :- C + A < B, !, fail.

triangle(A, B, C, _) :- (A = 0; B = 0; C = 0), !, fail.

triangle(S, S, S, "equilateral") :- !.

triangle(S, S, _, "isosceles") :- !.
triangle(S, _, S, "isosceles") :- !.
triangle(_, S, S, "isosceles") :- !.

triangle(A, B, C, "scalene") :- A \= B, B \= C, C \= A, !.